File: dndacademyBE/src/main/java/com/giulia/dndacademy/service/QuestionService.java
package com.giulia.dndacademy.service;

import com.giulia.dndacademy.dto.*;

import java.util.List;

public interface QuestionService {

    QuestionDTO createQuestion(CreateQuestionRequest request);

    List<QuestionDTO> getQuestionsByQuiz(Long quizId);

    QuestionDTO updateQuestion(Long questionId, UpdateQuestionRequest request);

    void deleteQuestion(Long questionId);

    List<AdminQuestionDTO> getAdminQuestionsByQuiz(Long quizId);

    AdminQuestionDTO createAnswer(Long questionId, CreateAnswerRequest request);

    AdminQuestionDTO updateAnswer(
            Long questionId,
            Long answerId,
            CreateAnswerRequest request
    );

    AdminQuestionDTO deleteAnswer(Long questionId, Long answerId);
}

File: dndacademyfe/src/services/question.service.ts
import { apiFetch } from "@/services/api"
import type {
  AdminQuestion,
  CreateAnswerRequest,
  CreateQuestionRequest,
  UpdateQuestionRequest,
} from "@/types/quiz"

export const questionService = {
  getAdminQuestionsByQuiz(quizId: number) {
    return apiFetch<AdminQuestion[]>(`/api/questions/admin/quiz/${quizId}`, {
      method: "GET",
      auth: true,
    })
  },

  createQuestion(request: CreateQuestionRequest) {
    return apiFetch<AdminQuestion>("/api/questions", {
      method: "POST",
      auth: true,
      body: JSON.stringify(request),
    })
  },

  updateQuestion(questionId: number, request: UpdateQuestionRequest) {
    return apiFetch<AdminQuestion>(`/api/questions/${questionId}`, {
      method: "PUT",
      auth: true,
      body: JSON.stringify(request),
    })
  },

  deleteQuestion(questionId: number) {
    return apiFetch<void>(`/api/questions/${questionId}`, {
      method: "DELETE",
      auth: true,
    })
  },

  createAnswer(questionId: number, request: CreateAnswerRequest) {
    return apiFetch<AdminQuestion>(`/api/questions/${questionId}/answers`, {
      method: "POST",
      auth: true,
      body: JSON.stringify(request),
    })
  },

  updateAnswer(
    questionId: number,
    answerId: number,
    request: CreateAnswerRequest,
  ) {
    return apiFetch<AdminQuestion>(
      `/api/questions/${questionId}/answers/${answerId}`,
      {
        method: "PUT",
        auth: true,
        body: JSON.stringify(request),
      },
    )
  },

  deleteAnswer(questionId: number, answerId: number) {
    return apiFetch<AdminQuestion>(
      `/api/questions/${questionId}/answers/${answerId}`,
      {
        method: "DELETE",
        auth: true,
      },
    )
  },
}
